StackExchange.Redis/StackExchange/Redis/RawResult.cs

```diff
@@ -117,81 +122,93 @@ internal RedisValue AsRedisValue()
-                {
-                    if (actual64[i] != expected64[i]) return false;
-                }
-                int index = count - spare;
-                while (spare-- != 0)
-                {
-                    if (actual8[index] != expected8[index]) return false;
-                }
+                offset += from.Length;
             }
             return true;
         }
 
         internal bool AssertStarts(byte[] expected)
         {
             if (expected == null) throw new ArgumentNullException(nameof(expected));
-            if (expected.Length > count) return false;
-            var actual = arr as byte[];
-            if (actual == null) return false;
+            if (expected.Length > _payload.Length) return false;
 
-            for (int i = 0; i < expected.Length; i++)
+            var rangeToCheck = _payload.Slice(0, expected.Length);
+            if (rangeToCheck.IsSingleSegment) return rangeToCheck.First.Span.SequenceEqual(expected);
+
+            int offset = 0;
+            foreach(var segment in rangeToCheck)
             {
-                if (expected[i] != actual[offset + i]) return false;
+                var from = segment.Span;
+                var to = new Span<byte>(expected, offset, from.Length);
+                if (!from.SequenceEqual(to)) return false;
+
+                offset += from.Length;
             }
             return true;
         }
 
         internal byte[] GetBlob()
         {
-            var src = (byte[])arr;
-            if (src == null) return null;
+            if (IsNull) return null;
 
-            if (count == 0) return emptyBlob;
+            if (_payload.IsEmpty) return Array.Empty<byte>();
 
-            byte[] copy = new byte[count];
-            Buffer.BlockCopy(src, offset, copy, 0, count);
-            return copy;
+            return _payload.ToArray();
         }
 
         internal bool GetBoolean()
         {
-            if (count != 1) throw new InvalidCastException();
-            byte[] actual = arr as byte[];
-            if (actual == null) throw new InvalidCastException();
-            switch (actual[offset])
+            if (_payload.Length != 1) throw new InvalidCastException();
+            switch (_payload.First.Span[0])
             {
                 case (byte)'1': return true;
                 case (byte)'0': return false;
                 default: throw new InvalidCastException();
             }
         }
 
-        internal RawResult[] GetItems()
+        internal ReadOnlySpan<RawResult> GetItems()
         {
-            return (RawResult[])arr;
+            if (Type == ResultType.MultiBulk)
+                return new ReadOnlySpan<RawResult>(_itemsOversized, 0, _itemsCount);
+            throw new InvalidOperationException();
         }
 
         internal RedisKey[] GetItemsAsKeys()
         {
-            RawResult[] items = GetItems();
-            if (items == null)
+            var items = GetItems();
+            if (IsNull)
```

Comment:
Seeing this pattern of GetItems() followed by an `IsNull` check after manifested in many places after some reactoring - worth a look globally

Reply:
it is intentional; the `GetItems()` includes a type check (multi-bulk, else exception); but other types can be null (well, ok - BulkString); if I inverted this to defer the fetch, I'd be losing this check. It is only a span, anyhows - no alloc.